### PR TITLE
fix(stream): canonicalise named_parameters() for streamed models

### LINKED ISSUE #369: Streamed model: named_parameters() still carries the wrapper's .inner. segment, and it caused a correctness gate to pass on an empty comparison
## What

v0.72.1 fixed adapter serialisation by making `StreamedDecoderLayer.state_dict()` delegate at the wrapper's own prefix. That fix was **serialisation-only, by design**, so `named_parameters()` on a streamed model still yields keys like:

```
base_model.model.model.layers.0.inner.self_attn.q_proj.lora_A.weight
```

while every non-streamed consumer expects the canonical spelling without `.inner.`.

## Why it is worth its own issue rather than staying a documented quirk

It has now demonstrably caused a **false green**. The first run of the #331 repair gate at 32B reported `grads exact 0/0` and `variant2_gradients_all_exact: true`, because the streamed and resident key sets did not intersect and `n_exact == n_compared` is trivially satisfied by `0 == 0` (`benchmarks/gate-h100-validation.md` L2627–2643). What exposed it was **the control arm not breaking**, not the green verdict.

That is the third time in that session a control caught a false positive, and the only one of the three whose root cause is a shipped API returning non-canonical names.

It is also the reason `--resume` into a streamed model was refused in the first place: `state_dict()` and `named_parameters()` disagree in memory.

## Fix path

Either of:

- **(A)** Make `named_parameters()` / `named_modules()` canonical the same way `state_dict()` is, and re-verify the v0.72.0 bit-exactness gates (they were deliberately left un-re-run last time because the change was serialisation-only; this one is not).
- **(B)** Keep the current shape and ship a documented `canonical_named_parameters(model)` helper plus a hard failure — not a warning — when a comparison between a streamed and a non-streamed model produces an empty key intersection.

(B) is cheaper and preserves the "gates stay valid" property; (A) removes the whole class. Whoever takes it should decide explicitly rather than drifting into (B) by default.

## Acceptance criteria

- [ ] A test that a streamed and a resident model built from the same checkpoint produce **intersecting** parameter-name sets, with a control that the assertion can fail
- [ ] Any comparison helper raises on an empty intersection rather than reporting `0/0` as success
- [ ] If (A): the v0.72.0/.2/.3 bit-exactness suites re-run and re-recorded
- [ ] The `--resume` refusal is revisited or its reason updated, since it cites this disagreement

### PR DESCRIPTION
**Pull request by AmirF194**
## Root cause

v0.72.1 made `StreamedDecoderLayer.state_dict()` canonical for serialisation, delegating at the wrapper's own prefix so a saved adapter loads into any normal model. `named_parameters()`/`named_modules()` were left untouched: every wrapped decoder layer's parameters still carry an extra `.inner.` segment (`...layers.0.inner.self_attn.q_proj...`) that a resident (non-streamed) model of the same checkpoint does not have.

That gap first looked harmless because nothing shipped compared the two by name, until the #331 repair gate at 32B did exactly that for gradients and read `grads exact 0/0` as success (`benchmarks/gate-h100-validation.md` L2627-2643): the streamed and resident key sets did not intersect, so `n_exact == n_compared` was vacuously true on nothing compared.

## The invariant

A name-keyed comparison between a streamed and a non-streamed model of the same checkpoint must be over parameter names that actually mean the same weight, and a comparison helper must fail loudly on an empty intersection rather than report it as a pass.

## Fix

Ships fix shape (B) from the issue (the cheaper one, preserving the "gates stay valid" property without re-running the v0.72.0 bit-exactness suites):

- `canonical_named_parameters(model)`: strips `.inner.` the same way `state_dict()` already does.
- `assert_canonical_parameters_intersect(model_a, model_b)`: computes both canonical name sets and raises `ValueError` if the intersection is empty, returning the shared set on success.

Fix shape (A), making `named_parameters()` itself canonical, would also require re-running the v0.72.0/.2/.3 bit-exactness suites on real hardware. This sandbox has no GPU, so I could not do that safely; (B) touches neither the forward path nor `state_dict()`, so those gates stay valid unexercised, the same way they did for the v0.72.1 fix itself.

## Verification

- New regression tests (`tests/test_issue369_canonical_named_parameters.py`) build a real streamed DPO wrapper and a resident model from the same tiny checkpoint (CPU, `python:3.12-slim`, `pip install -e ".[dev]"`). A control test proves the bug: the two models' raw LoRA parameter names share nothing (reproducing the gate's `0/0`). The other tests fail on `main` (`ImportError`, the functions do not exist yet) and pass on this branch: canonical names strip `.inner.`, the streamed and resident canonical sets are equal, and `assert_canonical_parameters_intersect` both raises on a disjoint pair and returns the shared names on an intersecting one.
- `ruff check src/soup_cli/ tests/` and `ruff format --check` on the changed files are clean.
- Full `tests/test_v07200.py` through `tests/test_v07204.py` and `tests/test_v07300.py` (460 tests, the existing layer-streaming and preference-loss suites): all pass, 19 skipped (CUDA-gated, expected offline).
- `coverage run --source=soup_cli.utils.layer_stream_runtime`: every changed line, including the raise branch, is covered by the new tests.
- Not verified: `mypy` errors in this sandbox on an unrelated numpy stub issue before reaching any changed code (pre-existing in this environment, and non-blocking in this repo's own CI). No GPU is available here, so nothing about the v0.72.0/.2/.3 bit-exactness gates was re-run; this fix does not touch the code paths they cover.

Fixes #369

### CODE REVIEW

**Review by MakazhanAlpamys — CHANGES_REQUESTED** [review]
Ran this rather than reading it, and every mechanical claim in the description holds.

## Verified

**Red/green reproduces exactly.** Your suite against `main`'s source (branch tests, main code):

```
4 failed, 1 passed
```

and the one that passes is the CONTROL. That is the form that means something — a control documents the bug, so it has to be green *before* the fix, and the accusations have to be red. On the branch: `5 passed`.

**No regressions.** `test_v07200` through `test_v07204` plus `test_v07300`: `468 passed, 6 skipped` here. `ruff check src/soup_cli/ tests/` clean. `tests/test_cli_startup_is_light.py` 23 passed — worth stating explicitly, since `layer_stream_runtime` is reachable from the light CLI path and neither new function imports torch.

**CI had never run, and that was not your doing.** The workflow sat in `action_required` because it is your first PR here, which is why the checks tab looked empty. Approved it: **11/11 green**, all nine matrix cells plus lint and type-check.

**(B) over (A) is the right call, and it is the call the issue asked for.** #369 says whoever takes it should "decide explicitly rather than drifting into (B) by default". You decided explicitly and wrote down why. I would have made the same one: (A) puts the forward path in scope, and re-earning the v0.72.0/.2/.3 bit-exactness gates needs hardware neither of us has here.

**One objection of mine, withdrawn in your favour.** My first note said `name.replace(".inner.", ".")` is a blind string edit where `state_dict()` delegates structurally. Then I found `_layer_name_map` (`layer_stream_runtime.py:1245`) doing `pname.replace(".base_layer.", ".")` for PEFT's wrapper segment — same idiom, same file, already shipped. It is the file's convention, not an improvisation.

## Blocking, and it is one line

**`test_returns_the_shared_canonical_names_when_they_intersect` passes with canonicalisation entirely disabled.** I mutated `canonical_named_parameters` into a pass-through:

```
$ # yield name, param        (the .replace removed)
FAILED ...::test_canonical_names_strip_the_inner_segment
FAILED ...::test_streamed_and_resident_canonical_names_intersect
2 failed, 3 passed
```

The helper's own positive test is among the three that survive. It is satisfied by `embed_tokens` / `norm` / `lm_head` — the weights that never carried `.inner.` — so `shared` comes back non-empty and `all(".inner." not in name ...)` holds trivially over them.

That is the #331 condition exactly: every LoRA name disjoint, and a non-empty set of irrelevant names reported as agreement. A test for a fix against false-greens should not be greenable by the false-green.

You already knew the mechanism — your control says so ("Non-layer weights like the embeddings already match without canonicalising") and is correctly scoped to `lora_` because of it. The scoping just did not carry into the positive test. `assert any("lora_" in name for name in shared)` closes it, and my mutation then fails there too.

The other mutation is clean, for contrast: making the empty-intersection branch not raise fails exactly `test_raises_on_empty_intersection` and nothing else (`1 failed, 4 passed`).

## Also before merge: the issue's fourth criterion

#369 asks that "the `--resume` refusal is revisited or its reason updated, since it cites this disagreement". It is still there, in the file this PR edits — `layer_stream_runtime.py:723-726`:

> `named_parameters()` still shows `.inner.`, i.e. loading INTO a streamed model stays unsupported (`--resume` is refused; the checkpoint/resume slot is v0.72.3).

v0.72.3 shipped that slot. `commands/train.py:534-541` already explains that resume works, through the load-side pre-hook that mirrors the save-side delegation. So the comment contradicts the code two files over, and it is the comment a future reader hits first because it sits on `state_dict()`. The first half of the sentence stays true — `named_parameters()` *is* still non-canonical, which is the reason your helper exists — the parenthetical is what went stale.

## Two notes, neither blocking

- `"inner"` is now spelled out in a fourth place (`__init__`, `_redirect_canonical_keys`, `__getattr__`, and your helper). A module-level constant would make it one copy. Pre-existing and not yours to fix here.
- `named_modules()` and `named_buffers()` carry the same segment. The issue's acceptance criteria are about parameter names, so shipping parameters only is in scope — but the CHANGELOG entry says `named_parameters()`/`named_modules()` while only the first is covered. Narrow the sentence and it matches the diff.

## One thing that is not about your PR

The comparison that produced `grads exact 0/0` lives in the harness behind `benchmarks/gate-h100-validation.md`, and that harness is not in the repo (#379). So the new helper has no in-tree caller, and nothing forces a future comparison through it. That is not a gap in this change — (B) is written that way — but it means #369's second criterion is met by *availability*, not by *enforcement*.

I am adding it to #379 as an acceptance item: when the harness is published, its parameter comparison goes through `assert_canonical_parameters_intersect`. Recording it here so it does not evaporate when this issue closes, which is exactly how the first false green survived.

---

Two edits and I merge. Both are one-liners, and the CI is already green, so nothing else is outstanding.

**Comment by AmirF194** [comment]
Both fixed.

`test_returns_the_shared_canonical_names_when_they_intersect` now asserts `any("lora_" in name for name in shared)`. Verified with the same mutation you ran: neutralizing the `.replace(".inner.", ".")` call now fails this test (1 failed, 1 passed) where it previously passed alongside the other assertion; full file suite is 5 passed otherwise.

Updated the stale comment at `state_dict()` too: it named `--resume` as refused, when v0.72.3's load-side pre-hook in `train.py` already covers it.

Also rebased past `main` for the CHANGELOG conflict with #353/#351.

**Comment by MakazhanAlpamys** [comment]
Merged. Both items verified by running them, not by reading the summary.

**The false green is closed.** Same mutation as before — `canonical_named_parameters` neutered to a pass-through:

```
before your fix:  2 failed, 3 passed   (the helper's own positive test survived)
after:            3 failed, 2 passed
                  + TestAssertCanonicalParametersIntersect::test_returns_the_shared_canonical_names_when_they_intersect
```

That third failure is the point of the round. Clean run on your branch: `5 passed`.

**The stale comment came back better than what I asked for.** I asked you to correct a false parenthetical. You rewrote the paragraph so it now says why `canonical_named_parameters()` exists and points at the load-side pre-hook that makes `--resume` work. The next person to read `state_dict()` gets the actual shape of the design instead of a fact that expired one release ago.

CI green on `0a3d5f9` — all nine cells plus lint and type-check. It sat unapproved again for the same reason as the first push (first-time contributor); that stops now that this is merged.

One thing I took myself rather than sending you round a third time for one word: the CHANGELOG entry says `named_parameters()`/`named_modules()` where only parameters are covered. Narrowed on `main`, referencing you.

#369 closes with all four criteria met, including the fourth, which is the one that usually goes missing when a fix lands as a helper.

You are in [`CONTRIBUTORS.md`](https://github.com/MakazhanAlpamys/Soup/blob/main/CONTRIBUTORS.md).

---

Worth naming what made this a good first PR here, since it is not the diff — 43 lines of source.

You found the failure through **the control not breaking**, and you built the regression test the same way round: a control that reproduces the bug and must be green before the fix, with the accusations red. That is the form this repo asks for and it is the part people usually skip. Then, offered a choice between the cheap fix and the thorough one, you took the cheap one and **wrote down why** — which is exactly what #369 asked for and what stops (B) from becoming the default by drift.

If you want another, these are adjacent to what you have been reading:

- **#348** — `calibrated_logits_bytes_per_element` is dead code; it is never wired into `estimate_stream_peak_vram`, so the opt-in calibration it exists for cannot take effect.
- **#347** — an explicit override for a refused VRAM pre-flight (`training.stream_vram_override`). The pre-flight over-predicts (#327) and there is currently no way past it.

Neither needs a GPU. No obligation, and thanks for this one.

**Comment by AmirF194** [comment]
Appreciate the detailed writeup, and thanks for the CHANGELOG fix and the CONTRIBUTORS.md add. Will take a look at #348 and #347.

**Comment by MakazhanAlpamys** [comment]
Both marked as claimed in the issues so nobody duplicates you — [#347](https://github.com/MakazhanAlpamys/Soup/issues/347#issuecomment-5280572236) and [#348](https://github.com/MakazhanAlpamys/Soup/issues/348#issuecomment-5280576013). GitHub will not let me set the assignee badge for a non-collaborator, so the comment is the marker.

Each carries context that was not in the issue body when it was filed, and one of them changes the shape of the fix:

**#347 turns out to block something outside the issue tracker.** There is a free-tier Colab notebook planned — reader caps a T4 to ~4 GB with `set_per_process_memory_fraction(0.25)` and watches a streamed 8B run agree with a resident reference. It cannot work today: `stream_setup.py:481` reads free VRAM with `torch.cuda.mem_get_info()`, a **device-level** driver query, and the allocator cap does not change what the driver reports. The pre-flight sees ~15 GB on a 16 GB card regardless.

So the override has to be able to **lower** the budget, not only raise it — a number, not a `force: true`. A boolean would close #347 as written and leave the notebook impossible. Suggested acceptance test in the comment: with the override set below real free VRAM, a config that fits the card must be *refused*.

**#348** — the guardrail is that it is not a route around #327's decision. The constant stays at 14 (14 under-predicts 0 of 10 measured rows; 12.311 under-predicts 10 of 10, and on Windows an under-prediction spills silently instead of raising). What is dead is the **opt-in, upward-only** calibration, and the hole it was written to catch — a future stack growing a fourth fp32 buffer in the loss path, under-budgeted by 12.5% with nothing to notice — is still open.

If you want an order: **#347 first**, it has the outside dependency.

No timeline expected on either. If one turns out bigger than it looks, say so and take the other — or neither, that is fine too.
